design/0001-metallb-crd.md

```diff
@@ -0,0 +1,228 @@
+# MetalLB Custom Resource Controller
+
+<!-- toc -->
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+- [Proposal](#proposal)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Design Overview](#design-overview)
+    - [Option1](#option1)
+    - [Option2](#option2)  
+  - [Configuration Flow](#configuration-flow)
+  - [CRD Extendability](#crd-extendability)
+- [Testing](#testing)
+- [Debugging and Troubleshooting](#debugging-and-troubleshooting)
+
+<!-- /toc -->
+
+## Summary
+
+The purpose of this enhancement is to propose an alternative configuration
+interface for MetalLB based on Kubernetes Custom Resource Definitions (CRDs)
+while also retaining support for the existing ConfigMap interface.
+
+## Motivation
+
+Today MetalLB allows only configuration using ConfigMap objects,
+which is simple from configuration perspective and doesn't
+require cluster-admin rights, However it doesn't have validation on the API
+server level, no support for API versioning and no fine-grained permission
+model via RBAC.
+Custom resource definition "CRD" are managed like any other resource,
+through the Kubernetes API, and eventually stored in the backend store Etcd.
+
+### Goals
+
+- Configuring MetalLB using kubernetes custom resource "CRD".
+- Retain support of the existing ConfigMap interface for backwards
+  compatibility.
+
+## Proposal
+
+Using [Kubebuilder](https://github.com/kubernetes-sigs/kubebuilder)
+framework to build new CRDs "MetalLB" objects and a controllers which will construct
+Metallb's Config objects, the Metallb Controller will watch those objects and use
+them to configure Metallb.
+
+### Risks and Mitigations
+
+Users need to either use CRD or ConfigMap to configure MetalLB
+but not both, if adminstrator mistakenly uses both, the CRD configuration will
+overwrite the user defined ConfigMap.
+
+## Design Details
+
+### Design Overview
+
+[Metallb CRD Design diagram](https://excalidraw.com/#json=5697710665498624,2uxfWYNvvkI6IqP-VEaQng)
+
+![Metallb CRD Design diagram](./metallb-crd.png)
+
+There are multiple proposals been discussed during the review of this PR
+
+#### Option1
+
+This proposal will implement the following CRD objects :-
+
+- Node selectors CRD which is a general objects to specify the node(s) of interest
+
+```
+node-selector:
+    - match-labels:
+    - match-expressions:
+```    
+
+- Protocol can now be layer2 or bgp but with FRR capability we can add ospf/is-is and other protocols
+
+```
+protocol:
```

Comment:
I don' understand how this will work. Why would this not be part of the AddressPools object?